File: ios/AlphaCityStampTour/AlphaCityStampTour/Network/DTO/StoreDTO.swift
//
//  StoreDTO.swift
//  AlphaCityStampTour
//

import Foundation

struct StoreRegisterRequest: Encodable {
    let name: String
    let category: String
    let ownerName: String
    let phone: String
    var address: String?
    var addressDetail: String?
    var description: String?
    var imageUrl: String?
    var storeCode: String?
    var operatingDays: String?
    var openTime: String?
    var closeTime: String?
}

struct StoreData: Decodable, Identifiable {
    let id: Int
    let name: String
    let category: String
    let ownerName: String
    let phone: String
    let address: String?
    let addressDetail: String?
    let description: String?
    let imageUrl: String?
    let storeCode: String?
    let operatingDays: String?
    let openTime: String?
    let closeTime: String?
    let latitude: Double?
    let longitude: Double?
    let status: String
    let mission: StoreMissionInfo?
    let storeCoupons: [StoreCouponDetail]?
    let program: StoreProgramInfo?
}

struct StoreMissionInfo: Decodable {
    let id: Int
    let name: String
    let type: String
}

struct StoreCouponDetail: Decodable {
    let coupon: StoreCouponData
}

struct StoreCouponData: Decodable {
    let id: Int
    let name: String
    let description: String?
    let imageUrl: String?
}

struct StoreProgramInfo: Decodable {
    let id: Int
    let name: String
}

// StoreData → ProgramData 변환 (ProgramDetailView 재사용)
extension StoreData {
    func toProgramData() -> ProgramData {
        let hours = [openTime, closeTime].compactMap { $0 }.joined(separator: " - ")
        let fullAddress = [address, addressDetail].compactMap { $0 }.joined(separator: " ")
        return ProgramData(
            id: -id,
            festivalId: nil,
            name: name,
            description: description,
            category: "food",
            subcategory: nil,
            hasCoupon: !(storeCoupons?.isEmpty ?? true),
            imageUrl: imageUrl,
            operatingHours: hours.isEmpty ? nil : hours,
            location: fullAddress.isEmpty ? nil : fullAddress,
            latitude: latitude,
            longitude: longitude,
            phone: phone,
            speaker: nil,
            ownerName: ownerName,
            storeCode: storeCode,
            operatingDays: operatingDays,
            startDate: "",
            endDate: "",
            status: "in_progress",
            events: nil,
            storeCoupons: storeCoupons?.map {
                StoreCouponInfo(storeId: id, storeName: name, couponId: $0.coupon.id, couponName: $0.coupon.name, couponDescription: $0.coupon.description)
            },
            stores: nil
        )
    }
}

File: android/app/src/main/java/com/alphacity/stamptour/network/dto/StoreDto.kt
package com.alphacity.stamptour.network.dto

import kotlinx.serialization.Serializable

@Serializable
data class StoreRegisterRequest(
    val name: String,
    val category: String,
    val ownerName: String,
    val phone: String,
    val address: String? = null,
    val addressDetail: String? = null,
    val description: String? = null,
    val imageUrl: String? = null,
    val storeCode: String? = null,
    val operatingDays: String? = null,
    val openTime: String? = null,
    val closeTime: String? = null,
)

@Serializable
data class StoreData(
    val id: Int,
    val name: String,
    val category: String,
    val ownerName: String,
    val phone: String,
    val address: String? = null,
    val addressDetail: String? = null,
    val description: String? = null,
    val imageUrl: String? = null,
    val storeCode: String? = null,
    val operatingDays: String? = null,
    val openTime: String? = null,
    val closeTime: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val status: String,
    val mission: StoreMissionInfo? = null,
    val storeCoupons: List<StoreCouponDetail>? = null,
    val program: StoreProgramInfo? = null,
)

@Serializable
data class StoreMissionInfo(
    val id: Int,
    val name: String,
    val type: String,
)

@Serializable
data class StoreCouponDetail(
    val coupon: StoreCouponData,
)

@Serializable
data class StoreCouponData(
    val id: Int,
    val name: String,
    val description: String? = null,
    val imageUrl: String? = null,
)

@Serializable
data class StoreProgramInfo(
    val id: Int,
    val name: String,
)

// StoreData → ProgramItem 변환 (ProgramDetailScreen 재사용)
fun StoreData.toProgramItem(): ProgramItem {
    val hours = if (openTime != null && closeTime != null) "$openTime - $closeTime" else null
    val fullAddress = listOfNotNull(address, addressDetail).joinToString(" ")
    return ProgramItem(
        id = -id,
        name = name,
        description = description,
        category = "food",
        imageUrl = imageUrl,
        operatingHours = hours,
        location = fullAddress.ifBlank { null },
        latitude = latitude,
        longitude = longitude,
        startDate = "",
        endDate = "",
        status = "in_progress",
        phone = phone,
        ownerName = ownerName,
        storeCode = storeCode,
        operatingDays = operatingDays,
        storeCoupons = storeCoupons?.map {
            StoreCouponInfo(storeId = id, storeName = name, couponId = it.coupon.id, couponName = it.coupon.name, couponDescription = it.coupon.description)
        },
    )
}
